nucypher/characters/control/controllers.py

```diff
@@ -173,41 +176,225 @@ def __init__(self,
+        return NotImplemented
+
+    @abstractmethod
+    def test_client(self):
+        return NotImplemented
+
+
+class JSONRPCController(CharacterControlServer):
+
+    _emitter_class = JSONRPCStdoutEmitter
+
+    def start(self):
+        _transport = self.make_control_transport()
+        reactor.run()  # < ------ Blocking Call (Reactor)
+
+    def test_client(self) -> JSONRPCTestClient:
+        test_client = JSONRPCTestClient(rpc_controller=self)
+        return test_client
+
+    def make_control_transport(self):
+        transport = stdio.StandardIO(JSONRPCLineReceiver(rpc_controller=self))
+        return transport
+
+    def get_interface(self, name: str) -> Callable:
+
+        # Examine the controller interface
+        interfaces = inspect.getmembers(self._internal_controller,
+                                        predicate=inspect.ismethod)
+
+        # Generate a mapping of the public interfaces
+        interfaces = {i[0]: i[1] for i in interfaces if not i[0].startswith('_')}
+
+        try:
+            interface = interfaces[name]
+        except KeyError:
+            raise self.emitter.MethodNotFound
+
+        return interface
+
+    def handle_server_notification(self, notification_request) -> int:
+        pass
+
+    def handle_procedure_call(self, control_request, *args, **kwargs) -> int:
+
+        # Validate request and read request metadata
+        jsonrpc2 = control_request['jsonrpc']
+        if jsonrpc2 != '2.0':
+            raise self.emitter.InvalidRequest
+
+        request_id = control_request['id']
+
+        # Read the interface's signature metadata
+        method_name = control_request['method']
+        method_params = control_request.get('params', dict())  # optional
+
+        # Lookup the public interface
+        interface = self.get_interface(name=method_name)
+
+        # Call the internal interface | pipe to output
+        return interface(request=method_params, request_id=request_id, *args, **kwargs)  # < ------- INLET
+
+    def validate_request(self, request: dict):
+
+        #
+        # Phase 1 - Metadata
+        #
+
+        required_fields = {'jsonrpc', 'method'}
+        optional_fields = {'id', 'params'}
+        all_fields = required_fields | optional_fields
+
+        try:
+            input_fields = set(request.keys())
+        except AttributeError:
+            raise self.emitter.InvalidRequest
+
+        contains_required_fields = required_fields.issubset(input_fields)
+
+        unique_fields = all_fields - input_fields - optional_fields
+        contains_valid_fields = not bool(unique_fields)
+
+        is_valid = all((contains_required_fields,
+                        contains_valid_fields))
+
+        if not is_valid:
+            raise self.emitter.InvalidRequest
+
+        #
+        # Phase 2 - Content Type
+        #
+
+        method_name = request['method']
+
+        try:
+            int(method_name)  # must not be a number
```

Comment:
What if it's `'1.1'`? `'1blah'`?
In Python 3, strings seem to have a useful method `isidentifier()`. Is it what we need here?
```
>>> 'x'.isidentifier()
True
>>> 'x123'.isidentifier()
True
>>> '1'.isidentifier()
False
>>> '1.1'.isidentifier()
False
>>> '1blah'.isidentifier()
False
```

Reply:
Yeah - this bit is a bit awkward. : Hmm... `.isidentifier`... thinking: